File: app/src/main/java/com/bopr/android/smailer/ui/HistoryDetailsDialogFragment.java
package com.bopr.android.smailer.ui;

import android.annotation.SuppressLint;
import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.text.format.DateFormat;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.DialogFragment;
import androidx.fragment.app.FragmentActivity;

import com.bopr.android.smailer.PhoneEvent;
import com.bopr.android.smailer.R;

import static com.bopr.android.smailer.PhoneEvent.REASON_NUMBER_BLACKLISTED;
import static com.bopr.android.smailer.PhoneEvent.REASON_TEXT_BLACKLISTED;
import static com.bopr.android.smailer.PhoneEvent.REASON_TRIGGER_OFF;
import static com.bopr.android.smailer.util.ResourceUtil.eventDirectionImage;
import static com.bopr.android.smailer.util.ResourceUtil.eventStateImage;
import static com.bopr.android.smailer.util.ResourceUtil.eventStateText;
import static com.bopr.android.smailer.util.ResourceUtil.eventTypeImage;
import static com.bopr.android.smailer.util.ResourceUtil.eventTypeText;
import static com.bopr.android.smailer.util.TagFormatter.formatter;
import static com.bopr.android.smailer.util.Util.formatDuration;

/**
 * Log item details dialog.
 *
 * @author Boris Pronin (<a href="mailto:boprsoft.dev@gmail.com">boprsoft.dev@gmail.com</a>)
 */
public class HistoryDetailsDialogFragment extends DialogFragment {

    private PhoneEvent value;

    void showDialog(FragmentActivity activity) {
        show(activity.getSupportFragmentManager(), "log_details_dialog");
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setRetainInstance(true);
    }

    @Override
    public void onDestroyView() {
        /* avoiding disappear on rotation */
        if (getDialog() != null && getRetainInstance()) {
            getDialog().setDismissMessage(null);
        }
        super.onDestroyView();
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        Dialog dialog = getDialog();
        if (dialog == null) {
            @SuppressLint("InflateParams")
            View view = LayoutInflater.from(getContext()).inflate(R.layout.fragment_log_details, null, false);

            view.<ImageView>findViewById(R.id.image_event_type).setImageResource(eventTypeImage(value));
            view.<ImageView>findViewById(R.id.image_event_direction).setImageResource(eventDirectionImage(value));
            view.<TextView>findViewById(R.id.text_title).setText(value.getPhone());
            view.<TextView>findViewById(R.id.text_message).setText(formatMessage(value));
            view.<TextView>findViewById(R.id.text_time).setText(formatTime(value.getStartTime()));
            view.<ImageView>findViewById(R.id.image_event_result).setImageResource(eventStateImage(value));
            view.<TextView>findViewById(R.id.text_result).setText(eventStateText(value));
            view.<TextView>findViewById(R.id.text_result_reason).setText(formatReason(value));
            view.<TextView>findViewById(R.id.text_type_title).setText(eventTypeText(value));
            view.<TextView>findViewById(R.id.text_recipient).setText(value.getRecipient());

            dialog = new AlertDialog.Builder(requireContext())
                    .setView(view)
                    .setPositiveButton(R.string.close, new DialogInterface.OnClickListener() {

                        @Override
                        public void onClick(DialogInterface dialogInterface, int which) {
                            dialogInterface.cancel();
                        }
                    })
                    .create();
        }
        return dialog;
    }

    private CharSequence formatReason(PhoneEvent event) {
        if (event.getState() == PhoneEvent.STATE_IGNORED) {
            if ((event.getStateReason() & REASON_NUMBER_BLACKLISTED) != 0) {
                return "(" + getString(R.string.number_in_blacklist) + ")";
            } else if ((event.getStateReason() & REASON_TEXT_BLACKLISTED) != 0) {
                return "(" + getString(R.string.text_in_blacklist) + ")";
            } else if ((event.getStateReason() & REASON_TRIGGER_OFF) != 0) {
                return "(" + getString(R.string.trigger_off) + ")";
            }
        }
        return null;
    }

    public void setValue(PhoneEvent value) {
        this.value = value;
    }

    private CharSequence formatMessage(PhoneEvent event) {
        if (event.isSms()) {
            return event.getText();
        } else if (event.isMissed()) {
            return getString(R.string.you_had_missed_call);
        } else {
            int pattern;
            if (event.isIncoming()) {
                pattern = R.string.you_had_incoming_call;
            } else {
                pattern = R.string.you_had_outgoing_call;
            }
            return formatter(requireContext())
                    .pattern(pattern)
                    .put("duration", formatDuration(event.getCallDuration()))
                    .format();
        }
    }

//    private String formatRecipient(PhoneEvent event) {
//        return formatter(requireContext())
//                .pattern(R.string.recipient_device)
//                .put("recipient", event.getRecipient())
//                .format();
//    }

    private CharSequence formatTime(Long time) {
        return DateFormat.format(getString(R.string._time_pattern), time);
    }
}

File: app/src/main/kotlin/com/bopr/android/smailer/ui/HistoryDetailsDialogFragment.kt
package com.bopr.android.smailer.ui

import android.text.format.DateFormat
import android.view.LayoutInflater
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.bopr.android.smailer.R
import com.bopr.android.smailer.messenger.Event
import com.bopr.android.smailer.messenger.Event.Companion.FLAG_BYPASS_NUMBER_BLACKLISTED
import com.bopr.android.smailer.messenger.Event.Companion.FLAG_BYPASS_TEXT_BLACKLISTED
import com.bopr.android.smailer.messenger.Event.Companion.FLAG_BYPASS_TRIGGER_OFF
import com.bopr.android.smailer.messenger.ProcessState.Companion.STATE_IGNORED
import com.bopr.android.smailer.messenger.ProcessState.Companion.STATE_PROCESSED
import com.bopr.android.smailer.provider.telephony.PhoneCallData
import com.bopr.android.smailer.util.*

/**
 * Call history log item details dialog fragment.
 *
 * @author Boris Pronin ([boris280471@gmail.com](mailto:boris280471@gmail.com))
 */
class HistoryDetailsDialogFragment(private val event: Event) :
    BaseDialogFragment("log_details_dialog") {

    override fun onCreateDialogView(inflater: LayoutInflater, root: ViewGroup?): View {
        return inflater.inflate(R.layout.dialog_history_details, root, false).apply {
            findViewById<ImageView>(R.id.image_processing_result).setImageResource(
                messageStateImage(event.processState)
            )
            findViewById<TextView>(R.id.text_result).setText(messageStateText(event.processState))
            findViewById<TextView>(R.id.text_recipient).text = event.target
            findViewById<TextView>(R.id.text_result_time).run {
                visibility = if (event.processState == STATE_PROCESSED) VISIBLE else GONE
                text = formatProcessTime(event)
            }
            findViewById<ImageView>(R.id.image_explain_result).run {
                visibility = if (event.processState == STATE_IGNORED) VISIBLE else GONE
                setOnClickListener {
                    onExplainResult()
                }
            }

            (event.payload as? PhoneCallData)?.let {
                findViewById<ImageView>(R.id.image_phone_call_type).setImageResource(
                    phoneCallTypeImage(it)
                )
                findViewById<ImageView>(R.id.image_phone_call_direction).setImageResource(
                    phoneCallDirectionImage(it)
                )
                findViewById<TextView>(R.id.text_title).text = it.phone
                findViewById<TextView>(R.id.text_message).text = formatMessage(it)
                findViewById<TextView>(R.id.text_time).text = formatTime(it.startTime)
                findViewById<TextView>(R.id.text_type_title).setText(phoneCallTypeText(it))
            }
        }
    }

    private fun onExplainResult() {
        val msg = buildString {
            if (FLAG_BYPASS_NUMBER_BLACKLISTED in event.bypassFlags) {
                append(getString(R.string.number_in_blacklist))
                append("\n\n")
            }
            if (FLAG_BYPASS_TEXT_BLACKLISTED in event.bypassFlags) {
                append(getString(R.string.text_in_blacklist))
                append("\n\n")
            }
            if (FLAG_BYPASS_TRIGGER_OFF in event.bypassFlags) {
                append(getString(R.string.trigger_off))
                append("\n\n")
            }
        }

        MessageDialog(getString(R.string.ignored), msg).show(this)
    }

    private fun formatProcessTime(event: Event): CharSequence? {
        return event.processTime?.run {
            getString(R.string.processed_at, formatTime(this))
        }
    }

    private fun formatTime(time: Long): CharSequence {
        return DateFormat.format(getString(R.string._time_pattern), time)
    }

    private fun formatMessage(info: PhoneCallData): CharSequence? {
        return when {
            info.isSms ->
                info.text

            info.isMissed ->
                getString(R.string.you_had_missed_call)

            else -> {
                val pattern = if (info.isIncoming) {
                    R.string.you_had_incoming_call
                } else {
                    R.string.you_had_outgoing_call
                }
                getString(pattern, formatDuration(info.callDuration))
            }
        }
    }
}